### PR TITLE
Downgrade The TypeLoc in VarDecl to a TypeRepr

### PR DESCRIPTION
**Pull request by CodaFi**
TypeCheckPattern used to splat the interface type into this, and
different parts of the compiler would check one or the other.   There is
now one source of truth: The interface type.  The type repr is now just
a signal that the user has written an explicit type annotation on
a parameter.  For variables, we will eventually be able to just grab
this information from the parent pattern.

EDIT: This started as a simple refactor, now I'm just nuking the old stateful code around this that we don't need anymore.  There's diagnostic improvements from this.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -4027,10 +4027,7 @@ ParamSpecifierRequest::evaluate(Evaluator &evaluator,
   return ParamSpecifier::Default;
 }
 
-static void validateParameterType(ParamDecl *decl) {
-  if (auto ty = decl->getTypeLoc().getType())
-    return;
-
+static Type validateParameterType(ParamDecl *decl) {
```
Comment (CodaFi) [inline lib/Sema/TypeCheckDecl.cpp]: *Gee, this whole thing looks suspiciously requestifiable now*

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckDecl.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/TypeCheckDecl.cpp]: DOOOOO EEETTTT

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**